From 0cd97a08c03628d237a20d8709e6fc6be2624619 Mon Sep 17 00:00:00 2001
From: c4pp4
Date: Tue, 12 Nov 2019 13:07:11 +0100
Subject: [PATCH 1/1] online-accounts: enable passing data

Uncomment property override and use string array to convey panel arguments
instead of GVariant. It's used by gnome-calendar to launch new account
set up page.

Signed-off-by: c4pp4
---
 .../cc-online-accounts-panel.c                | 47 ++++---------------
 1 file changed, 8 insertions(+), 39 deletions(-)

diff --git a/panels/online-accounts/cc-online-accounts-panel.c b/panels/online-accounts/cc-online-accounts-panel.c
index 0479b62..694479c 100644
--- a/panels/online-accounts/cc-online-accounts-panel.c
+++ b/panels/online-accounts/cc-online-accounts-panel.c
@@ -309,32 +309,11 @@ sort_func (GtkListBoxRow *a,
 
 static void
 command_add (CcGoaPanel *panel,
-             GVariant   *parameters)
+             const gchar *provider_name)
 {
-  GVariant *v, *preseed = NULL;
   GoaProvider *provider = NULL;
-  const gchar *provider_name = NULL;
 
   g_assert (panel != NULL);
-  g_assert (parameters != NULL);
-
-  switch (g_variant_n_children (parameters))
-    {
-      case 3:
-        g_variant_get_child (parameters, 2, "v", &preseed);
-      case 2:
-        g_variant_get_child (parameters, 1, "v", &v);
-        if (g_variant_is_of_type (v, G_VARIANT_TYPE_STRING))
-          provider_name = g_variant_get_string (v, NULL);
-        else
-          g_warning ("Wrong type for the second argument (provider name) GVariant, expected 's' but got '%s'",
-                     (gchar *)g_variant_get_type (v));
-        g_variant_unref (v);
-        break;
-      default:
-        g_warning ("Unexpected parameters found, ignore request");
-        goto out;
-    }
 
   if (provider_name != NULL)
     {
@@ -345,12 +324,11 @@ command_add (CcGoaPanel *panel,
           goto out;
         }
 
-      add_account (panel, provider, preseed);
+      add_account (panel, provider, NULL);
     }
 
 out:
   g_clear_object (&provider);
-  g_clear_pointer (&preseed, g_variant_unref);
 }
 
 static void
@@ -363,26 +341,17 @@ cc_goa_panel_set_property (GObject *object,
     {
       case PROP_PARAMETERS:
         {
-          GVariant *parameters, *v;
+          gchar **parameters;
           const gchar *first_arg = NULL;
 
-          parameters = g_value_get_variant (value);
-          if (parameters == NULL)
+          parameters = g_value_get_boxed (value);
+          if (parameters == NULL || *parameters == '\0')
             return;
 
-          if (g_variant_n_children (parameters) > 0)
-            {
-                g_variant_get_child (parameters, 0, "v", &v);
-                if (g_variant_is_of_type (v, G_VARIANT_TYPE_STRING))
-                  first_arg = g_variant_get_string (v, NULL);
-                else
-                  g_warning ("Wrong type for the second argument GVariant, expected 's' but got '%s'",
-                             (gchar *)g_variant_get_type (v));
-                g_variant_unref (v);
-            }
+          first_arg = parameters[0];
 
           if (g_strcmp0 (first_arg, "add") == 0)
-            command_add (CC_GOA_PANEL (object), parameters);
+            command_add (CC_GOA_PANEL (object), parameters[1]);
           else if (first_arg != NULL)
             select_account_by_id (CC_GOA_PANEL (object), first_arg);
 
@@ -519,7 +488,7 @@ cc_goa_panel_class_init (CcGoaPanelClass *klass)
   object_class->constructed = cc_goa_panel_constructed;
   object_class->dispose = cc_goa_panel_dispose;
 
-  //g_object_class_override_property (object_class, PROP_PARAMETERS, "parameters");
+  g_object_class_override_property (object_class, PROP_PARAMETERS, "argv");
 
   gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/control-center/online-accounts/online-accounts.ui");
 
-- 
2.23.0

